File: nextcertification/public/js/sales_order.js

frappe.ui.form.on("Sales Order", {
     setup: function(frm) {
		frm.add_fetch("application", "registration_no", "registration_no");
	},

    refresh:function(frm){
        if(frm.doc.docstatus ==1){
            frm.add_custom_button(__('Application'),function() {
                return frappe.call({
                    method: 'nextcertification.sales_order.make_application',
                    args:{
                     doc_name:frm.doc.name
                    },
                    callback: function(r) {
                        var doc = frappe.model.sync(r.message);
                        frappe.set_route("Form", doc[0].doctype, doc[0].name);
                    }
                });
            }, __("Create"));
        }
	},

	"section" : function(frm) {
            frm.set_query('product_category', function(doc) {
                return {
                    filters: {'section': frm.doc.section}
                };
            });
        },
    "product_category" : function(frm) {
        frm.set_query('sub_category', function(doc) {
            return {
                filters: {'product_category': frm.doc.product_category}
            };
        });
    },
    "service_type" : function(frm){
       // console.log('-------- frm status doc ------',frm.doc.docstatus);
          frappe.call({
                method: "nextcertification.sales_order.check_registration_req",
                args: {
                    doctype: "Service Types",
                    name: frm.doc.service_type
                    },
                callback:function(r) {
                    if(r.message.registration == 1)
                    {
                    frm.set_df_property("registration_no", "hidden", 0);
                    frm.set_df_property("registration_no", "reqd", 1);

                    frm.refresh_field("registration_no")
                    }
                    else
                    {
                    frm.set_df_property("registration_no", "hidden", 1);
                    frm.refresh_field("registration_no")
                    }
                }
            })
        }


});

File: nextcertification/sales_order.py
import frappe
from frappe.model.document import Document

@frappe.whitelist()
def make_application(doc_name =None):
    if doc_name:
        doc = frappe.get_doc("Sales Order",doc_name)
        application =frappe.new_doc("Application")
        application.customer = doc.customer
        application.service_type = doc.service_type
        application.section = doc.section
        application.product_category = doc.product_category
        application.sub_category = doc.sub_category
        application.registration_no = doc.registration_no
        application.sales_order_id = doc.name
        return application

@frappe.whitelist()
def check_registration_req(name=None):
	if name:
		dic={}
		service = frappe.get_doc("Service Types",name)
		dic['certificate'] = service.is_certificate_req
		dic['registration'] = service.is_registration_req
		return dic
	return False


def add_sales_to_application(doc, method):
    if doc.application:
        app = frappe.get_doc('Application', doc.application)
        print("*************",app)
        if(app):
            app.sales_order_id = doc.name
            app.flags.ignore_validate_update_after_submit = True
            app.save(ignore_permissions=True)
